src/Microsoft.OData.Core/UriParser/Binders/InBinder.cs

```
@@ -97,21 +97,25 @@ private CollectionNode GetCollectionOperandFromToken(QueryToken queryToken, IEdm
             LiteralToken literalToken = queryToken as LiteralToken;
             if (literalToken != null)
             {
-                string originalLiteralText = literalToken.OriginalText;
-
                 // Parentheses-based collections are not standard JSON but bracket-based ones are.
                 // Temporarily switch our collection to bracket-based so that the JSON reader will
                 // correctly parse the collection. Then pass the original literal text to the token.
-                string bracketLiteralText = originalLiteralText;
-                if (bracketLiteralText[0] == '(')
+                string bracketLiteralText = literalToken.OriginalText;
+
+                if (bracketLiteralText[0] == '(' || bracketLiteralText[0] == '[')
                 {
-                    Debug.Assert(bracketLiteralText[bracketLiteralText.Length - 1] == ')',
-                        "Collection with opening '(' should have corresponding ')'");
+                    Debug.Assert((bracketLiteralText[0] == '(' && bracketLiteralText[^1] == ')') || (bracketLiteralText[0] == '[' && bracketLiteralText[^1] == ']'),

```

Can we get here in reality? Or is this designed just to assert our expectations during development? If we can get here in reality, we should throw an exception instead

---

We can't get here with `unbalanced brackets` in reality. The unbalanced brackets will be captured in the `ExpressionLexer` and throw an expression. The method that does that is [AdvanceThroughBalancedExpression](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs#L1209)

```cs
private void AdvanceThroughBalancedExpression(char startingCharacter, char endingCharacter)
{
    int currentBracketDepth = 1;

    while (currentBracketDepth > 0)
    {
        ...
        // the rest of code
        if (this.ch == null)
        {
            throw new ODataException(ODataErrorStrings.ExpressionLexer_UnbalancedBracketExpression); <=Exception thrown here
        }

        this.NextChar();
    }
}
```

The method severally:
- [AdvanceThroughBalancedExpression('(', ')')](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs#L433)
- [AdvanceThroughBalancedExpression('(', ')')](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs#L548)
- [AdvanceThroughBalancedExpression('{', '}')](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs#L647)
- [AdvanceThroughBalancedExpression('[', ']')](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs#L652)